airesources/OCaml/game.ml

```diff
@@ -0,0 +1,256 @@
+  state.all_ships
+;;
+
+let docked_ships planet =
+  planet.docked_ships
+;;
+
+let get_planet_docked_ship planet ship_id =
+  try
+    Some (List.find (fun ship ->
+      ship.s_entity.id = ship_id
+    ) planet.docked_ships)
+  with _ -> None
+;;
+
+let planet_is_owned planet =
+  planet.p_entity.owner >= 0
+;;
+
+let planet_is_full planet =
+  List.length planet.docked_ship_ids >= planet.num_docking_spots
+;;
+
+let rec positive_int_degrees a =
+  if a < 0 then positive_int_degrees (a + 360)
+  else a
+;;
+
+let cmd_thrust ship m a =
+  let magnitude = int_of_float m in
+  let angle = positive_int_degrees ((int_of_float (floor (a +. 0.5))) mod 360) in
+    `Thrust (ship.s_entity.id, magnitude, angle)
+;;
+
+let cmd_dock ship planet =
+  `Dock (ship.s_entity.id, planet.p_entity.id)
+;;
+
+let cmd_undock ship =
+  `Undock ship.s_entity.id
+;;
+
+let can_dock ship planet =
+  let distance = calculate_distance_between ship.s_entity planet.p_entity in
+    distance <= planet.p_entity.radius +. Const.dock_radius
+;;
+
+let position_entity x y =
+ {
+  id = (-1);
+  x = x;
+  y = y;
+  radius = 0.1;
+  health = 0;
+  owner = (-1)
+ }
+;;
+
+let intersect_segment_circle start eend circle fudge =
+  let dx = eend.x -. start.x in
+  let dy = eend.y -. start.y in
+  let a = dx ** 2. +. dy ** 2. in
+  let b = (-2.) *. 
+    (start.x ** 2. -. 
+      start.x *. eend.x -. 
+      start.x *. circle.x +. 
+      eend.x *. circle.x +. 
+     start.y ** 2. -. 
+     start.y *. eend.y -. 
+     start.y *. circle.y +. 
+     eend.y *. circle.y) 
+  in
+(* This variable seems to be unused. *)
+(*  let c = (start.x -. circle.x) ** 2. +. (start.y -. circle.y) ** 2. in *)
+    if a = 0. then (
+      calculate_distance_between start circle <= circle.radius +. fudge
+    )
+    else (
+      let t = min (-.b /. (2. *. a)) 1. in
+        if t < 0. then false
+        else (
+          let closest_x = start.x +. dx *. t in
+          let closest_y = start.y +. dy *. t in
+          let pos_ent = position_entity closest_x closest_y in
+          let closest_distance = calculate_distance_between pos_ent circle in
+            closest_distance <= circle.radius +. fudge
+        )
+    )
+  
+;;
+
+
+(* Note that unlike the Python starter, this obstacles_between function
+ * expects a list of potential obstacles, rather than a list of types
+ * to exclude. It is up to the calling function to filter the obstacles.
+ *)
+let obstacles_between e1 e2 obstacles =
+  List.filter (fun obst ->
+    (not ((obst == e1) || (obst == e2)))
+    && intersect_segment_circle e1 e2 obst (Const.ship_radius +. 0.1)
+  ) obstacles
+;;
+
+let direct_navigate ship distance max_speed angle =
+  let speed = if distance >= max_speed then max_speed else distance in
+    (* Not sure why adding 180 is necessary here, but it is. *)
+    (* Perhaps a disagreement about how radians are converted to degrees? *)
+    (* Or where the zero point is? *)
+    cmd_thrust ship speed (angle +. 180.)
```

Comment:
This makes me uncomfortable :-)
Why do you think it's necessary here?

Reply:
Now fixed. I had imported my rad2deg and deg2rad functions from another project without checking them properly. No more adding of 180 degrees.

I've also made the other changes suggested.